Source/Core/Core/HW/WiimoteReal/IOLinux.cpp

```diff
@@ -4,218 +4,315 @@
+{
+  std::string entries = Config::Get(Config::MAIN_WIIMOTE_AUTO_CONNECT_ADDRESSES);
+  for (auto& bt_address_str : SplitString(entries, ','))
+  {
+    const auto bt_addr = Common::StringToBluetoothAddress(bt_address_str);
+    if (!bt_addr.has_value())
+    {
+      WARN_LOG_FMT(WIIMOTE, "Bad Auto Connect Bluetooth Address: {}", bt_address_str);
+      continue;
+    }
+
+    Common::ToLower(&bt_address_str);
+    if (!IsNewWiimote(bt_address_str))
+      continue;
+
+    found_wiimotes.push_back(new WiimoteLinux(*bt_addr));
+    NOTICE_LOG_FMT(WIIMOTE, "Added Wiimote with fixed address ({}).", bt_address_str);
+  }
+}
+
+WiimoteScannerLinux::WiimoteScannerLinux()
 {
+  Open();
+}
+
+bool WiimoteScannerLinux::Open()
+{
+  if (IsReady())
+    return true;
+
   // Get the id of the first Bluetooth device.
   m_device_id = hci_get_route(nullptr);
   if (m_device_id < 0)
   {
-    NOTICE_LOG_FMT(WIIMOTE, "Bluetooth not found.");
-    return;
+    NOTICE_LOG_FMT(WIIMOTE, "Bluetooth not found. hci_get_route: {}", Common::LastStrerrorString());
+    return false;
   }
 
   // Create a socket to the device
   m_device_sock = hci_open_dev(m_device_id);
   if (m_device_sock < 0)
   {
-    ERROR_LOG_FMT(WIIMOTE, "Unable to open Bluetooth.");
-    return;
+    ERROR_LOG_FMT(WIIMOTE, "Unable to open Bluetooth. hci_open_dev: {}",
+                  Common::LastStrerrorString());
+    return false;
   }
+
+  m_is_device_open.store(true, std::memory_order_relaxed);
+  return true;
 }
 
 WiimoteScannerLinux::~WiimoteScannerLinux()
 {
-  if (IsReady())
-    close(m_device_sock);
+  Close();
+}
+
+void WiimoteScannerLinux::Close()
+{
+  if (!IsReady())
+    return;
+
+  m_is_device_open.store(false, std::memory_order_relaxed);
+
+  close(std::exchange(m_device_sock, -1));
+  m_device_id = -1;
 }
 
 bool WiimoteScannerLinux::IsReady() const
 {
-  return m_device_sock > 0;
+  return m_is_device_open.load(std::memory_order_relaxed);
 }
 
-void WiimoteScannerLinux::FindWiimotes(std::vector<Wiimote*>& found_wiimotes, Wiimote*& found_board)
+struct InquiryRequest : hci_inquiry_req
 {
-  WiimoteScannerLinux::AddAutoConnectAddresses(found_wiimotes);
+  static constexpr int MAX_INFOS = 255;
+  std::array<inquiry_info, MAX_INFOS> scan_infos;
+};
+static_assert(sizeof(InquiryRequest) ==
+              sizeof(hci_inquiry_req) + sizeof(inquiry_info) * InquiryRequest::MAX_INFOS);
+
+// Returns 0 on success or some error number on failure.
+static int HciInquiry(int device_socket, InquiryRequest* request)
+{
+  const auto done_event = UnixUtil::CreateEventFD(0, 0);
+  Common::ScopeGuard close_guard([&] { close(done_event); });
+
+  int hci_inquiry_errorno = 0;
+
+  // Unplugging a BT adapter causes `hci_inquiry` to block forever (inside `ioctl`).
+  // Fortunately it does produce a signal on the socket so we can poll for that.
+  // Performing the inquiry on thread lets us interrupt `ioctl` if the socket signals.
```

Comment:
I'm not gonna question this but this seems like bizarre design, is there even a way to handle that without killing the thread the ioctl runs on??

Reply:
I agree that it's terrible. I think it's a bug in the `HCIINQUIRY` `ioctl`. I should probably submit a bug report.

This is the only thing I could come up with that didn't just block forever.

Reply:
FYI, I think `pthread_cancel` secretly just sends a USR1 signal or something like that (which interrupts the `ioctl`), and because `pthread_testcancel` isn't used the thread just ends normally.

I don't think it's as crazy as "killing the thread".